servicenow/stackstate_checks/servicenow/servicenow.py

```diff
@@ -1,78 +1,124 @@
+except AttributeError:  # Python 2
+    json_parse_exception = ValueError
 
-# 3rd party
 import requests
+from schematics import Model
+from schematics.exceptions import DataError
+from schematics.types import URLType, StringType, ListType, IntType, DictType, DateTimeType, ModelType, BaseType
 
-# project
-from stackstate_checks.base import ConfigurationError, AgentCheck, TopologyInstance
+from stackstate_checks.base import AgentCheck, TopologyInstance, Identifiers
 from stackstate_checks.base.errors import CheckException
 
-# inbuilt
-import yaml
-
 BATCH_DEFAULT_SIZE = 2500
-EVENT_TYPE = SOURCE_TYPE_NAME = 'servicenow'
-
-
-def _get_mandatory_instance_values(instance):
-    """
-    Check if mandatory instance values are present
-    :param instance:
-    :return: Strings base_url, password, user
-    """
-    try:
-        user = instance['user']
-        password = instance['password']
-        base_url = instance['url']
-    except KeyError as e:
-        raise ConfigurationError('ServiceNow CMDB topology instance missing "{}" value.'.format(e))
-    return base_url, password, user
-
-
-class InstanceInfo:
-    def __init__(self, instance_tags, base_url, auth, sys_class_filter, batch_size, timeout):
-        self.instance_tags = instance_tags
-        self.base_url = base_url
-        self.auth = auth
-        self.sys_class_filter = sys_class_filter
-        self.batch_size = batch_size
-        self.timeout = timeout
+BATCH_MAX_SIZE = 10000
+TIMEOUT = 20
+CRS_BOOTSTRAP_DAYS_DEFAULT = 100
+CRS_DEFAULT_PROCESS_LIMIT = 1000
+
+
+class WrapperType(BaseType):
+    def __init__(self, field, value_mapping, **kwargs):
+        self.field = field(**kwargs)
+        self.value_mapping = value_mapping
+        super(WrapperType, self).__init__(**kwargs)
+
+    def convert(self, value, context=None):
+        if value is None:
+            return self.default()
+        if context.new:
+            try:
+                value = self.value_mapping(value)
+            except KeyError:
+                value = ''
+        return self.field.convert(value)
+
+    def export(self, value, format, context=None):
+        self.field.export(value, format, context)
+
+
+class ChangeRequest(Model):
+    number = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    state = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    cmdb_ci = DictType(StringType, required=True)
+    sys_updated_on = WrapperType(DateTimeType, value_mapping=lambda x: x['value'])
+    business_service = WrapperType(StringType, value_mapping=lambda x: x['display_value'])
+    service_offering = WrapperType(StringType, value_mapping=lambda x: x['display_value'])
+    short_description = WrapperType(StringType, required=True, value_mapping=lambda x: x['display_value'])
+    description = WrapperType(StringType, default='No description', value_mapping=lambda x: x['display_value'])
```

Comment:
default values are not working for `WrappedType`

Reply:
I'm not sure that required works. @zandrewitte I'll need to your help with this.

Reply:
We fixed that :)